scala_proto/private/scalapb_aspect.bzl

```diff
@@ -0,0 +1,267 @@
+        labels = {},
+        in_scalacopts = [],
+        print_compile_time = False,
+        expect_java_output = False,
+        scalac_jvm_flags = [],
+        scalac = ctx.attr._scalac,
+    )
+
+    return JavaInfo(
+        source_jar = scalapb_jar,
+        deps = deps_java_info + implicit_deps,
+        runtime_deps = deps_java_info + implicit_deps,
+        exports = deps_java_info + implicit_deps,
+        output_jar = output,
+        compile_jar = output,
+    )
+
+def _empty_java_info(deps_java_info, implicit_deps):
+    merged_deps = java_common.merge(deps_java_info + implicit_deps)
+    return java_common.create_provider(
+        use_ijar = False,
+        compile_time_jars = depset(transitive = [merged_deps.compile_jars]),
+        transitive_compile_time_jars = depset(
+            transitive = [merged_deps.transitive_compile_time_jars],
+        ),
+        transitive_runtime_jars = depset(
+            transitive = [merged_deps.transitive_runtime_jars],
+        ),
+    )
+
+####
+# This is applied to the DAG of thrift_librarys reachable from a deps
+# or a scalapb_scala_library. Each thrift_library will be one scalapb
+# invocation assuming it has some sources.
+def _scalapb_aspect_impl(target, ctx):
+    deps = [d[ScalaPBAspectInfo].java_info for d in ctx.rule.attr.deps]
+
+    if ProtoInfo not in target:
+        java_info = target[JavaInfo]
+        src_jars = depset()
+        outs = depset()
+        transitive_ti = merge_proto_infos(
+            [
+                d[ScalaPBAspectInfo].proto_info
+                for d in ctx.rule.attr.deps
+            ],
+        )
+    else:
+        target_ti = target[ProtoInfo]
+        transitive_ti = merge_proto_infos(
+            [
+                d[ScalaPBAspectInfo].proto_info
+                for d in ctx.rule.attr.deps
+            ] + [target_ti],
+        )
+        # we sort so the inputs are always the same for caching
+        compile_protos = sorted(target_ti.direct_sources)
+        transitive_protos = sorted(target_ti.transitive_sources)
+
+        toolchain = ctx.toolchains["@io_bazel_rules_scala//scala_proto:toolchain_type"]
+        flags = []
+        imps = [j[JavaInfo] for j in ctx.attr._implicit_compile_deps]
+
+        if toolchain.with_grpc:
+            flags.append("grpc")
+            imps.extend([j[JavaInfo] for j in ctx.attr._grpc_deps])
+
+        if toolchain.with_flat_package:
+            flags.append("flat_package")
+
+        if toolchain.with_single_line_to_string:
+            flags.append("single_line_to_proto_string")
+
+        # if toolchain.with_java:
+        #     flags.append("java_conversions")
+
+        if compile_protos:
+            # we sort so the inputs are always the same for caching
+            compile_proto_map = {}
+            for ct in compile_protos:
+                compile_proto_map[ct] = True
+            include_thrifts = sorted([
+                trans
+                for trans in compile_protos
```

Comment:
please forgive my python ignorance but when will trans not be in the map?
Two lines above we add every ct in compile_protos in the map and trans also comes from compile_protos

Reply:
Thats fair, let me ensure this is the case. This is likely just a degree of cargo culting from the scrooge aspect